File: backend/models/ha_identity.py
# Smart Home Health
# Copyright (C) 2026 John Carty
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""Home Assistant identities observed on ingress requests.

Every HA user who opens the add-on panel gets a row here (upserted on
/api/auth/ha/login), which is exactly the population an administrator wants
to link to app users — no Supervisor API privileges needed for a picker.
The actual user mapping lives on ``users.ha_user_id``.
"""
from datetime import datetime

from sqlalchemy import Column, DateTime, Integer, String

from db import Base


class HASeenIdentity(Base):
    """An HA user id we've seen arrive via ingress, with display metadata."""
    __tablename__ = "ha_seen_identities"

    id = Column(Integer, primary_key=True, index=True)
    ha_user_id = Column(String(32), unique=True, nullable=False, index=True)
    username = Column(String(100), nullable=True)       # HA username; absent for some auth providers
    display_name = Column(String(100), nullable=True)   # HA display name, user-editable
    first_seen = Column(DateTime, default=datetime.utcnow, nullable=False)
    last_seen = Column(DateTime, default=datetime.utcnow, nullable=False)

File: frontend/src/services/haIdentity.js
/*
 * Smart Home Health
 * Copyright (C) 2026 John Carty
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
// Home Assistant identity mapping (system-admin only endpoints).
import { apiFetch, API_BASE_URL } from '../config';

const base = () => `${API_BASE_URL}/api/auth/ha`;

async function asJsonOrThrow(res, fallback) {
  if (res.ok) return res.status === 204 ? null : res.json();
  let detail = fallback;
  try {
    const body = await res.json();
    if (typeof body?.detail === 'string') detail = body.detail;
  } catch { /* non-JSON error body */ }
  const err = new Error(detail);
  err.status = res.status;
  throw err;
}

export async function listHaIdentities() {
  return asJsonOrThrow(await apiFetch(`${base()}/identities`), 'Failed to load Home Assistant identities');
}

// Merged directory: everyone in HA (when running as the add-on) plus
// identities seen on ingress and mapped users. {available, users: [...]}.
export async function getHaDirectory() {
  return asJsonOrThrow(await apiFetch(`${base()}/directory`), 'Failed to load the Home Assistant user directory');
}

// Create a passwordless app user pre-linked to an HA identity.
export async function importHaUser({ ha_user_id, username, full_name, role_ids = [] }) {
  return asJsonOrThrow(
    await apiFetch(`${base()}/import`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ ha_user_id, username, full_name, role_ids }),
    }),
    'Failed to import Home Assistant user',
  );
}

export async function linkHaIdentity(haUserId, userId) {
  return asJsonOrThrow(
    await apiFetch(`${base()}/identities/${haUserId}/link`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ user_id: userId }),
    }),
    'Failed to link identity',
  );
}

export async function unlinkHaIdentity(haUserId) {
  return asJsonOrThrow(
    await apiFetch(`${base()}/identities/${haUserId}/link`, { method: 'DELETE' }),
    'Failed to unlink identity',
  );
}

export async function forgetHaIdentity(haUserId) {
  return asJsonOrThrow(
    await apiFetch(`${base()}/identities/${haUserId}`, { method: 'DELETE' }),
    'Failed to remove identity',
  );
}
